source/simulator.hpp

```diff
@@ -10,42 +15,174 @@
 #include "device/sender.hpp"
 #include "device/switch.hpp"
 #include "link.hpp"
+#include "logger/logger.hpp"
+#include "scheduler.hpp"
 
 namespace sim {
 
+using routing_table_t =
+    std::unordered_map<std::shared_ptr<IRoutingDevice>, std::shared_ptr<ILink>>;
+
+static routing_table_t bfs(std::shared_ptr<IRoutingDevice>& start_device) {
+    routing_table_t routing_table;
```

Comment:
Let's do `bfs` a Simutator's private method and try to use `std::variant` here (I hope it will wokr as I think):
```suggestion
using Device = std::variant<TSwitch, TSender, TReceiver>;
static routing_table_t bfs(std::shared_ptr<Device> start_device) {
    routing_table_t routing_table;
```